### PR TITLE
remote: Rework reconnection code

### PR DESCRIPTION
**Pull request by jansol**
Take two on #1504

### CODE REVIEW

File: lib/CL/pocl_threads.h [hunk 1]
```diff
@@ -45,4 +47,14 @@
 
 #endif
 
+#define POCL_IGNORE_SIGNAL_IN_THREAD(s)                                       \
```
Comment (franz) [inline lib/CL/pocl_threads.h:50]: Could you move this macro into `pocl_threads_c.h`

File: pocld/snakeoil.key [hunk 2]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (franz) [inline pocld/snakeoil.key:1]: What is the purpose of the key/cert ?

File: pocld/virtual_cl_context.cc [hunk 3]
```diff
@@ -141,14 +142,17 @@ class VirtualCLContext : public VirtualContextBase {
...
+      MainThread.join();
     POCL_MSG_PRINT_GENERAL("VCTX: DEST\n");
 
     // Wake up IO threads in case they were waiting for a connection
     WriteFast->setConnection(nullptr);
     WriteSlow->setConnection(nullptr);
 
     // make sure no shared context tries to broadcast stuff
-    std::unique_lock<std::mutex> lock(main_mutex);
+    std::unique_lock<std::mutex> lock(MainMutex);
```
Comment (pjaaskel) [inline pocld/virtual_cl_context.cc:155]:
```suggestion
    std::unique_lock<std::mutex> Lock(MainMutex);
```

**Review by franz — CHANGES_REQUESTED** [review]

File: pocld/virtual_cl_context.cc [hunk 4]
```diff
@@ -329,9 +334,9 @@ void VirtualCLContext::nonQueuedPush(Request *req) {
   POCL_MSG_PRINT_GENERAL("VCTX NON-QUEUED PUSH (msg: %" PRIu64 ")\n",
                          uint64_t(req->Body.msg_id));
 
-  std::unique_lock<std::mutex> lock(main_mutex);
-  main_que.push_back(req);
-  main_cond.notify_one();
+  std::unique_lock<std::mutex> lock(MainMutex);
```
Comment (pjaaskel) [inline pocld/virtual_cl_context.cc:337]:
```suggestion
  std::unique_lock<std::mutex> Lock(MainMutex);
```

File: pocld/virtual_cl_context.cc [hunk 5]
```diff
@@ -358,7 +363,7 @@ void VirtualCLContext::requestExit(int code, const char *reason) {
 }
 
 void VirtualCLContext::broadcastToPeers(const Request &req) {
-  std::unique_lock<std::mutex> lock(main_mutex);
+  std::unique_lock<std::mutex> lock(MainMutex);
```
Comment (pjaaskel) [inline pocld/virtual_cl_context.cc:366]:
```suggestion
  std::unique_lock<std::mutex> Lock(MainMutex);
```

File: pocld/virtual_cl_context.cc [hunk 6]
```diff
@@ -415,10 +420,10 @@ int VirtualCLContext::run() {
       return e;
     }
 
-    std::unique_lock<std::mutex> lock(main_mutex);
-    if (main_que.size() > 0) {
-      Request *request = main_que.front();
-      main_que.pop_front();
+    std::unique_lock<std::mutex> lock(MainMutex);
```
Comment (pjaaskel) [inline pocld/virtual_cl_context.cc:423]:
```suggestion
    std::unique_lock<std::mutex> Lock(MainMutex);
```

**Review by pjaaskel — APPROVED** [review]
Can be done post pull.

File: pocld/snakeoil.key (hunk 2 above)
Comment (pjaaskel) [reply pocld/snakeoil.key:1]: I assumed he had a test that uses these, but seems not.

Comment (jansol) [reply pocld/snakeoil.key]: Oh, yeah those were lying around from some unrelated stuff and weren't supposed to get picked up

**Review by franz — APPROVED** [review]

**Pull request merged**